ethpm_cli/install.py

```
@@ -0,0 +1,104 @@
+    if (config.ethpm_dir / pkg.alias).is_dir():
+        raise InstallError(f"{pkg.alias} is already installed.")
+
+    # Create temporary package directory
+    tmp_pkg_dir = Path(tempfile.mkdtemp())
+    install_tmp_package(pkg, tmp_pkg_dir, config.ipfs_backend)
+
+    # Copy temp pacakge directory to ethpm dir namespace
+    dest_pkg_dir = config.ethpm_dir / pkg.alias
+    validate_parent_directory(config.ethpm_dir, dest_pkg_dir)
+    shutil.copytree(tmp_pkg_dir, dest_pkg_dir)
+    update_ethpm_lock(pkg, (config.ethpm_dir / "ethpm.lock"))
+
+
+def install_tmp_package(
+    pkg: Package, tmp_pkg_dir: Path, ipfs_backend: BaseIPFSBackend
+) -> None:
+    (tmp_pkg_dir / "manifest.json").touch()
+    (tmp_pkg_dir / "manifest.json").write_bytes(pkg.raw_manifest)
+
+    write_sources_to_disk(pkg, tmp_pkg_dir, ipfs_backend)
+    write_build_deps_to_disk(pkg, tmp_pkg_dir, ipfs_backend)
+    tmp_ethpm_lock = tmp_pkg_dir.parent / "ethpm.lock"
+    update_ethpm_lock(pkg, tmp_ethpm_lock)
+
+
+def write_sources_to_disk(
+    pkg: Package, pkg_dir: Path, ipfs_backend: BaseIPFSBackend
+) -> None:
+    for path, source in pkg.manifest["sources"].items():
+        if is_ipfs_uri(source):
+            source_contents = to_text(ipfs_backend.fetch_uri_contents(source)).rstrip(
+                "\n"
+            )
+        else:
+            # for inlined sources
+            source_contents = source
+        target_file = pkg_dir / "src" / path
+        target_dir = target_file.parent
+        if not target_dir.is_dir():
+            target_dir.mkdir(parents=True)
+        target_file.touch()
+        validate_parent_directory((pkg_dir / "src"), target_file)
+        target_file.write_text(source_contents)
+
+
+def write_build_deps_to_disk(
+    pkg: Package, pkg_dir: Path, ipfs_backend: BaseIPFSBackend
+) -> None:
+    if "build_dependencies" in pkg.manifest:
+        child_ethpm_dir = pkg_dir / "ethpm_packages"
+        if not child_ethpm_dir.is_dir():
```

I think the next line will fail loudly if there happens to be a file in this location but that file is not a directory.  Is this intentional?

---

If I understand right, when we're writing to the tmpdir we don't have to really worry about that happening? In fact, we can remove the check on line 86 since if a manifest is valid, there will never be build dependencies with conflicting names.

---

your call.